lib/iris/coords.py

```diff
@@ -52,6 +52,702 @@
+        #: have no meaning to Iris.
+        self.attributes = attributes
+
+        # Set up DataManager attributes and values.
+        self._values_dm = None
+        self._values = values
+
+    def __getitem__(self, keys):
+        """
+        Returns a new dimensional metadata whose values are obtained by
+        conventional array indexing.
+
+        .. note::
+
+            Indexing of a circular coordinate results in a non-circular
+            coordinate if the overall shape of the coordinate changes after
+            indexing.
+
+        """
+        # Fetch the values.
+        values = self._values_dm.core_data()
+
+        # Index values with the keys.
+        _, values = iris.util._slice_data_with_keys(
+            values, keys)
+
+        # Copy values after indexing to avoid making metadata that is a
+        # view on another metadata.
+        values = values.copy()
+
+        # If the metadata is a coordinate and it has bounds, repeat the above
+        # with the bounds.
+        if isinstance(self, Coord):
+            if self.has_bounds():
+                bounds = self._bounds_dm.core_data()
+                _, bounds = iris.util._slice_data_with_keys(bounds, keys)
+                bounds = bounds.copy()
+            else:
+                bounds = None
+
+            copy_args = dict(points=values, bounds=bounds)
+        else:
+            copy_args = dict(values=values)
+
+        # The new metadata is a copy of the old one with replaced content.
+        new_metadata = self.copy(**copy_args)
+
+        return new_metadata
+
+    def copy(self, values=None):
+        """
+        Returns a copy of this dimensional metadata object.
+
+        Kwargs:
+
+        * values
+            An array of values for the new dimensional metadata object.
+            This may be a different shape to the orginal values array being
+            copied.
+
+        """
+        new_metadata = copy.deepcopy(self)
+        if values is not None:
+            new_metadata._values_dm = None
+            new_metadata._values = values
+
+        return new_metadata
+
+    def _sanitise_array(self, src, ndmin):
+        if _lazy.is_lazy_data(src):
+            # Lazy data : just ensure ndmin requirement.
+            ndims_missing = ndmin - src.ndim
+            if ndims_missing <= 0:
+                result = src
+            else:
+                extended_shape = tuple([1] * ndims_missing + list(src.shape))
+                result = src.reshape(extended_shape)
+        else:
+            # Real data : a few more things to do in this case.
+            # Ensure the array is writeable.
+            # NB. Returns the *same object* if src is already writeable.
+            result = np.require(src, requirements='W')
+            # Ensure the array has enough dimensions.
+            # NB. Returns the *same object* if result.ndim >= ndmin
+            # Coords do not preserve masks of input points.
```

Comment:
As raised in #3419, coordinates for some reason throw away the mask. I kept that behaviour in case it was intentional. But if we decided it is a "bug" I could fix it here?

Reply:
Just noticed this.  As noted [here in #3419](https://github.com/SciTools/iris/issues/3419#issue-498188507), masked (aux) coordinates definitely are _allowed_ in CF.
So I think it *is* a bug.